src/blueapi/preprocessors/attach_metadata.py

```diff
@@ -32,10 +33,9 @@ def attach_metadata(
         Iterator[Msg]: Plan messages
     """
     yield from bps.wait_for([provider.update])
-    directory_info = provider()
-    yield from bpp.inject_md_wrapper(
-        plan, md={DATA_SESSION: directory_info.filename_prefix}
-    )
+    directory_info: DirectoryInfo = provider()
+    yield from bpp.inject_md_wrapper(plan, md={DATA_SESSION: directory_info.prefix})
+    provider.clear()
```

Comment:
Won't this mean that detectors subsequently calling `provider()` will try to write to `None`?

Reply:
Yes, I think the issue is actually that update() should always fetch a new instance rather than only the first time.

Reply:
It was intended to be a caching thing, where `update()` would generate a brand new scan number and save location, so you would end up with this:

![image](https://github.com/DiamondLightSource/blueapi/assets/29771545/405a78db-5951-4d01-b16a-4daf61a72a94)

If it fetches a new instance every time, wouldn't the detectors each write to a different scan number, which would be different in turn from the resource doc?

Reply:
Yes, you're right (and update() doesn't hold onto the previous one as I thought it did in the previous comment), I was confused by the change to no longer swallow the Exception in update() [as required merge into Bluesky had occured], which previously made sure the prior collection was None but no longer does! :zap: